### PR TITLE
Fixed #23879 -- Replaced conditional discovery by test skipping

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

File: django/contrib/postgres/operations.py [hunk 1]
```diff
@@ -12,6 +12,9 @@ def state_forwards(self, app_label, state):
         pass
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        engine = schema_editor.connection.settings_dict['ENGINE']
+        if not ('postgresl' in engine or 'postgis' in engine):
```
Comment (MarkusH) [inline django/contrib/postgres/operations.py]: apart from the missing `q` in `postgresql`: I'm -1 on this change. If I have a custom engine inheriting from the built-in PG engine this change forces me to name it appropriately. Can't we inspect the actual engine class and see the vendor?

Comment (claudep) [inline django/contrib/postgres/operations.py]: Yes, absolutely. This patch is still not ready, I'm working on some failures.

File: tests/postgres_tests/fields.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+"""
+from django.db import models
+
+try:
+    from django.contrib.postgres.fields import (
+        ArrayField, BigIntegerRangeField, DateRangeField, DateTimeRangeField,
+        FloatRangeField, HStoreField, IntegerRangeField,
+    )
+except ImportError:
+    class DummyArrayField(models.Field):
```
Comment (carljm) [inline tests/postgres_tests/fields.py]: is this necessary to even allow the import of the migration file, or could it also just be a mock (like how the extension operations are handled)?

**Comment by carljm** [comment]
This looks great! :+1:

**Comment by claudep** [comment]
Thanks for the review! I'll revisit the fields part to see if I can use mocking (not sure). I'm a but busy right now.

File: tests/postgres_tests/fields.py (hunk 2 above)
Comment (claudep) [inline tests/postgres_tests/fields.py]: Apparently, a simple `mock` doesn't do the job. I didn't try hard, though, but maybe some reviewer wants to try some other strategy?

Comment (carljm) [inline tests/postgres_tests/fields.py]: I think this is fine as-is.

**Comment by claudep** [comment]
I just added some docs for the `required_db_features`/`required_db_vendor` new `Meta` attributes.

File: docs/ref/models/options.txt [hunk 3]
```diff
@@ -283,6 +283,31 @@ Django quotes column and table names behind the scenes.
...
     a :ref:`proxy model <proxy-models>`.
 
+``required_db_features``
+------------------------
+
+.. attribute:: Options.required_db_features
+
+    .. versionadded:: 1.9
+
+    List of database features that the current connection should have so as the
```
Comment (timgraham) [inline docs/ref/models/options.txt]: so that the

File: docs/ref/models/options.txt [hunk 4]
```diff
@@ -283,6 +283,31 @@ Django quotes column and table names behind the scenes.
...
+    model is considered during the migration phase. For example, if you set this
+    list to ``['gis_enabled']``, the model will only be synchronized on
+    GIS-enabled databases. Also useful to skip some models when testing with
```
Comment (timgraham) [inline docs/ref/models/options.txt]: It's also useful...

File: docs/ref/models/options.txt [hunk 5]
```diff
@@ -283,6 +283,31 @@ Django quotes column and table names behind the scenes.
...
+    several database backends.
+
+``required_db_vendor``
+------------------------
```
Comment (timgraham) [inline docs/ref/models/options.txt]: shorter underline

File: tests/gis_tests/distapp/models.py [hunk 6]
```diff
@@ -12,6 +12,7 @@ class NamedModel(models.Model):
 
     class Meta:
         abstract = True
+        required_db_features = ('gis_enabled',)
```
Comment (timgraham) [inline tests/gis_tests/distapp/models.py]: Might be nice to use a list to be consistency with docs.

File: tests/postgres_tests/models.py [hunk 7]
```diff
@@ -1,39 +1,43 @@
-from django.contrib.postgres.fields import (
-    ArrayField, BigIntegerRangeField, DateRangeField, DateTimeRangeField,
-    FloatRangeField, HStoreField, IntegerRangeField,
-)
 from django.db import connection, models
 
+from .fields import *  # NOQA
```
Comment (timgraham) [inline tests/postgres_tests/models.py]: do we need import *? It seems like the same names can be imported regardless of vendor, but maybe I missed something (in which case an explanatory comment would be helpful)

**Comment by timgraham** [comment]
Did you consider whether or not we need to add the new meta options to `AlterModelOptions.ALTER_OPTION_KEYS`?

Should we add the options to the release notes?

**Comment by timgraham** [comment]
I cannot start the tests on a system without GIS dependencies installed:
```
Testing against Django installed in '/home/tim/django/django'
Traceback (most recent call last):
  File "tests/runtests.py", line 429, in <module>
    options.debug_sql)
  File "tests/runtests.py", line 228, in django_tests
    state = setup(verbosity, test_labels)
  File "tests/runtests.py", line 207, in setup
    apps.set_installed_apps(settings.INSTALLED_APPS)
  File "/home/tim/django/django/apps/registry.py", line 313, in set_installed_apps
    self.populate(installed)
  File "/home/tim/django/django/apps/registry.py", line 108, in populate
    app_config.import_models(all_models)
  File "/home/tim/django/django/apps/config.py", line 198, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib/python3.4/importlib/__init__.py", line 109, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 2231, in _gcd_import
  File "<frozen importlib._bootstrap>", line 2214, in _find_and_load
  File "<frozen importlib._bootstrap>", line 2203, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1448, in exec_module
  File "<frozen importlib._bootstrap>", line 321, in _call_with_frames_removed
  File "/home/tim/django/tests/gis_tests/relatedapp/models.py", line 1, in <module>
    from django.contrib.gis.db import models
  File "/home/tim/django/django/contrib/gis/db/models/__init__.py", line 12, in <module>
    "See also https://docs.djangoproject.com/en/%s/ref/contrib/gis/install/geolibs/" % get_docs_version())
django.core.exceptions.ImproperlyConfigured: GEOS is required and has not been detected. Are you sure it is installed? See also https://docs.djangoproject.com/en/dev/ref/contrib/gis/install/geolibs/
```

**Comment by claudep** [comment]
Tim, I hopefully fixed most of your comments. I'm not sure about `ALTER_OPTION_KEYS`, I'm struggling understanding the consequences... And for release notes, maybe, but I think it mostly affects Django itself.

**Comment by timgraham** [comment]
For `ALTER_OPTION_KEYS`, it basically controls whether or not the autodetector will generate a migration if that value changes. For example, we don't have index/unique_together in there because that's a separate operation. I think the answer is yes, we do want changes in those values to be reflected in migrations.

I had a similar thought about the release notes (in fact, I wasn't sure you were going to document the flags at all, or just keep them for internal use). If you don't have any other changes, I think this is fine for commit.

**Pull request merged**